File: leetcode/easy/235-lca-bst/iter.py
"""
  @author Vortexx2
  Problem 235 - LCA of BST

  Runtime - 76 ms
  Memory Usage - 18.1 MB
"""
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, x):
    self.val = x
    self.left = None
    self.right = None


class Solution:
  def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
    minNode, maxNode = None, None
    if p.val < q.val:
      minNode, maxNode = p, q
    else:
      minNode, maxNode = q, p

    while root.val < minNode.val or root.val > maxNode.val:
      if root.val < minNode.val:
        root = root.right

      else:
        root = root.left

    return root

File: leetcode/easy/235-lca-bst/iter.cpp
/**
 * @file iter.cpp
 * @author Vortexx2
 * @brief Problem 235 - LCA of BST
 * @date 25-08-2021
 * 
 * Runtime - 24 ms O(logn)
 * Memory Usage - 23.3 MB O(1)
 */
#include <iostream>

using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    TreeNode *minNode, *maxNode;
    if (p->val < q->val)
      minNode = p, maxNode = q;
    else
      minNode = q, maxNode = p;

    while (root->val < minNode->val || root->val > maxNode->val) {
      if (root->val < minNode->val)
        root = root->right;
      else
        root = root->left;
    }

    return root;
  }
};